maro/simulator/scenarios/supply_chain/units/consumer.py

```diff
@@ -51,46 +46,99 @@ def __init__(
 
         self._unit_order_cost: float = 0
 
+        self._init_statistics()
+
+    def _init_statistics(self) -> None:
+        self.order_quantity_on_the_way: Dict[int, int] = defaultdict(int)
+        self._open_orders: Counter = Counter()
+
+        # Dynamically changing statistics.
+        self.pending_scheduled_order_quantity: int = 0
+        self.in_transit_quantity: int = 0  # In transit = pending scheduled + on the way.
+
+        # Only incremental action valid.
+        self._total_order_num: int = 0
+        self._finished_order_num: int = 0
+        self._expired_order_num: int = 0
+        self._actual_order_leading_time: Counter = Counter()
+        self._order_schedule_delay_time: Counter = Counter()
+
+    @property
+    def order_statistics(self) -> Dict[str, Union[int, Counter]]:
+        return {
+            # The accumulated order number that have been placed to distribution unit, = finished + expired + active.
+            "total_order_num": self._total_order_num,
+            # The accumulated finished order number (expired order number not included).
+            "finished_order_num": self._finished_order_num,
+            # The accumulated expired order number (expired due to waiting too long to load by upstream).
+            "expired_order_num": self._expired_order_num,
+            # The accumulated actual leading time distribution, leading time = finished time - order creation time.
+            "actual_order_leading_time": self._actual_order_leading_time,
+            # The accumulated scheduling delayed time distribution, delay time = departure time - order creation time.
+            "order_schedule_delay_time": self._order_schedule_delay_time,
+            # The current total pending scheduled (waiting to load by upstream) product quantity.
+            "pending_scheduled_quantity": self.pending_scheduled_order_quantity,
+            # The current active ordered product quantity, active = pending scheduled + on the way (+ pending unload).
+            "active_ordered_quantity": self.in_transit_quantity,
+            # The product quantity that will be received in a future time window.
+            # The ones that not scheduled yet are not included here.
+            "expected_future_received": self.get_pending_order_daily(),
+        }
+
     def get_pending_order_daily(self, tick: int) -> List[int]:
         ret = [
             self.order_quantity_on_the_way[tick + i] for i in range(self.world.configs.settings["pending_order_len"])
         ]
-        if tick in self.order_quantity_on_the_way:  # Remove data at tick to save storage
-            self.order_quantity_on_the_way.pop(tick)
         return ret
 
-    @property
-    def in_transit_quantity(self) -> int:
-        return self._in_transit_quantity
+    def on_order_successfully_placed(self, order: Order) -> None:
```

Comment:
I remember we had a discussion about this naming issue: a method called `on_XXX` should be bound with an event handler, and only be called when the handler is triggered. It should not be called actively. Similar to the following several methods.

Reply:
rename to `handle_xxx` instead